packages/plugin-figma-blade-coverage/src/main.ts

```diff
@@ -0,0 +1,429 @@
+        // callback to stopTraversal for children of a node
+        // true: we shall stop
+        // false: we shall keep traversing children
+        if (!traversedNode.visible) {
+          return true;
+        }
+
+        if (
+          traversedNode.type === 'INSTANCE' &&
+          (BLADE_COMPONENT_IDS.includes(
+            (traversedNode.mainComponent?.parent as ComponentSetNode)?.key ?? '',
+          ) ||
+            BLADE_COMPONENT_IDS.includes(traversedNode.mainComponent?.key ?? ''))
+        ) {
+          // we shall stop traversal further if we have found that an instance is Blade instance
+          // if we keep traversing then chances are the metrics will be skewed because Blade components are composed of non-blade themselves
+          // in code analytics we can add "data-*" to all the children till leaf nodes but over here we can't hence we stop
+          return true;
+        }
+        return false;
+      },
+    );
+  } catch (error: unknown) {
+    console.error(error);
+    figma.notify('⚠️ Error in rendering coverage card. Please try again', { error: true });
+    figma.closePlugin();
+  }
+
+  return {
+    bladeComponents,
+    bladeTextStyles,
+    bladeColorStyles,
+    nonBladeComponents,
+    nonBladeTextStyles,
+    nonBladeColorStyles,
+    totalLayers,
+  };
+};
+
+const getPageMainFrameNodes = (nodes: SceneNode[]): SceneNode[] => {
+  const mainFrameNodes: SceneNode[] = [];
+  try {
+    for (const node of nodes) {
+      if (getParentNode(node)?.type === 'PAGE') {
+        // if selection is top level frame then start the coverage count
+        // await calculateCoverage(node);
+        mainFrameNodes.push(node);
+      } else {
+        // if the selection is not the top level frame then traverse up till we find the frame and then start the coverage count
+        // await calculateCoverage(mainFrameNode);
+        mainFrameNodes.push(traverseUpTillMainFrame(node) as SceneNode);
+      }
+    }
+  } catch (error: unknown) {
+    console.error(error);
+    figma.notify('⚠️ Error in identifying main frame node. Please try again', { error: true });
+    figma.closePlugin();
+  }
+
+  return mainFrameNodes;
+};
+
+const main = async (): Promise<void> => {
+  // plugin used
+  // const pluginUsageCount = await incrementTotalUseCountAsync();
+
+  // const analytics = AnalyticsBrowser.load({ writeKey: '6lpfX5loXnTbBFVo2FbJHpjins0hGaC4' });
+  // await analytics.track('Blade Coverage Plugin Used', {
+  //   pluginUsageCount,
+  //   fileName: figma.currentPage.parent?.name,
+  //   pageName: figma.currentPage.name,
+  // });
+  figma.skipInvisibleInstanceChildren = true;
+  figma.notify('Calculating Coverage', { timeout: Infinity });
+
+  let nodes: SceneNode[] = [];
+  if (figma.currentPage.selection.length > 0) {
+    // you already have the selection, run the plugin
+    //@ts-expect-error type 'readonly SceneNode[]' is 'readonly' and cannot be assigned to the mutable type 'SceneNode[]'
```

Comment:
you can make the nodes readonly

```
  let nodes: readonly SceneNode[] = [];
```

Also make the getPageMainFrameNodes accept `nodes: readonly SceneNode[]`

Better to specify nodes are readonly so that we accidentally don't modify consumer's design.

Reply:
done